pkg/azuredisk/controllerserver.go

```diff
@@ -885,16 +885,66 @@ func (d *Driver) CreateSnapshot(ctx context.Context, req *csi.CreateSnapshotRequ
+	if location != d.cloud.Location {
+		snapshotNameLocal := azureutils.CreateValidDiskName("local_" + snapshotName)
+		if !incremental {
+			snapshotNameLocal = snapshotName
 		}
 
-		azureutils.SleepIfThrottled(rerr.Error(), azureconstants.SnapshotOpThrottlingSleepSec)
-		return nil, status.Error(codes.Internal, fmt.Sprintf("create snapshot error: %v", rerr.Error()))
+		klog.V(2).Infof("begin to create snapshot(%s, incremental: %v) under rg(%s) region(%s)", snapshotNameLocal, incremental, resourceGroup, d.cloud.Location)
+		if rerr := d.cloud.SnapshotsClient.CreateOrUpdate(ctx, subsID, resourceGroup, snapshotNameLocal, snapshot); rerr != nil {
+			if strings.Contains(rerr.Error().Error(), "existing disk") {
+				return nil, status.Error(codes.AlreadyExists, fmt.Sprintf("request snapshot(%s) under rg(%s) already exists, but the SourceVolumeId is different, error details: %v", snapshotNameLocal, resourceGroup, rerr.Error()))
+			}
+
+			azureutils.SleepIfThrottled(rerr.Error(), azureconstants.SnapshotOpThrottlingSleepSec)
+			return nil, status.Error(codes.Internal, fmt.Sprintf("create snapshot error: %v", rerr.Error()))
+		}
+		klog.V(2).Infof("create snapshot(%s) under rg(%s) region(%s) successfully", snapshotNameLocal, resourceGroup, d.cloud.Location)
+
+		csiSnapshotLocal, err := d.getSnapshotByID(ctx, subsID, resourceGroup, snapshotNameLocal, sourceVolumeID)
+		if err != nil {
+			return nil, err
+		}
+
+		klog.V(2).Infof("create snapshot location(%s) is different from disk location(%s)", location, d.cloud.Location)
+		if incremental {
+			klog.V(2).Infof("begin to copy an incremental snapshot from region(%s) to new region(%s)", d.cloud.Location, location)
+			copySnapshot := snapshot
+			copySnapshot.SnapshotProperties.CreationData.SourceResourceID = &csiSnapshotLocal.SnapshotId
+			copySnapshot.SnapshotProperties.CreationData.CreateOption = compute.CopyStart
+			copySnapshot.Location = &location
+
+			klog.V(2).Infof("begin to create snapshot(%s, incremental: %v) under rg(%s) region(%s)", snapshotName, incremental, resourceGroup, location)
+			if rerr := d.cloud.SnapshotsClient.CreateOrUpdate(ctx, subsID, resourceGroup, snapshotName, copySnapshot); rerr != nil {
+				if strings.Contains(rerr.Error().Error(), "existing disk") {
+					return nil, status.Error(codes.AlreadyExists, fmt.Sprintf("request snapshot(%s) under rg(%s) already exists, but the SourceVolumeId is different, error details: %v", snapshotName, resourceGroup, rerr.Error()))
+				}
+
+				azureutils.SleepIfThrottled(rerr.Error(), azureconstants.SnapshotOpThrottlingSleepSec)
+				return nil, status.Error(codes.Internal, fmt.Sprintf("create snapshot error: %v", rerr.Error()))
+			}
+			klog.V(2).Infof("create snapshot(%s) under rg(%s) region(%s) successfully", snapshotName, resourceGroup, location)
```

Comment:
It takes a non-trivial amount of time for the snapshot to be copied to the destination region, and this copy could fail. We should monitor the copy progress as described in [Copy an incremental snapshot to a new region](https://learn.microsoft.com/en-us/azure/virtual-machines/disks-copy-incremental-snapshot-across-regions?tabs=azure-cli) and report status.